bfdd: protocol refactory

rzalamena opened the pull request:
### Summary

This PR contains a `bfdd` code refactory to fix bugs and make the protocol more RFC 5880 friendly.

 Highlights:

  * Move more control code out of `bfd_packet.c`;
  * Don't do a poll sequence for echo interval changes;
  * Don't ever send a BFD control packet with poll and final bit set together;
  * Centralize timer handling code into a single place;
  * Slow down control packets on connection setup (down -> init -> before up);
  * Slow down control packets on connection loss;
  * Generate random identifiers instead of using sequential numbers;
  * Fix `bfdd` start up with configuration containing interfaces (it broke when `bfdd` started using zebra to learn about interfaces);
  * Other small misc fixes;


### Components

`bfdd`

rzalamena commented:
CI:rerun

pguibert6WIND commented:
overall ok, the functional is here.
however, I wanted to clarify two points:
- the vrf support. when it is mentioend that single hop does not need to have vrf config, I don't think it is good idea to have different configuration attributes when configuring single hop or multi hop.
to go further, when bfd will be in yang, do you mean single hop and multihop will not be nested on the same yang data model? ( under vrf)?
- the way the bfd config is saved, while interface or vrf is not yet ready. Something more generic could be though. for instance, call vrf_get() for vrf creation, by using vrf name. ( I wrote something on https://github.com/FRRouting/frr/issues/3741 ).
- the 3rd point (random generation) is not blocking.

rzalamena commented:
@pguibert6WIND I think you are ahead of me in the VRF area and thank you for the input. I would say there was no planning in getting VRF for BFD yet, but the issue you pointed out is an important resource.

Addressing your points:

  1. '... * o go further, when bfd will be in yang, do you mean single hop and multihop will not be nested on the same yang data model? ( under vrf)? '

I don't know there is no yang model as far as I know, but since there is no working VRF implementation feel free to suggest one design (and CLI changes). At the moment there is no `vrf` blocks and selecting interfaces is enough to figure out which VRF we are operating (that is why `vrf` option only applies to `multihop`, because it has no effect on single hop).

  2. 'Something more generic could be though. for instance, call vrf_get() for vrf creation, by using vrf name.'

I agree, to fix the interface loading problem I had to implement the observer pattern to keep track when they start to exist. Let's keep this discussion inside the issue you've mentioned so we don't lose it.

This PR is mostly a refactory with no new features. I'd leave the VRF discussion to an issue or another PR.

pguibert6WIND commented:
ok, lets go for keeping the observer in bfd for now.

rzalamena commented on bfdd/bfdd_vty.c:
```diff
@@ -131,6 +131,10 @@ DEFUN_NOSH(
+		vty_out(vty, "%% VRF only applies with multihop.\n");
+		return CMD_WARNING_CONFIG_FAILED;
+	}
 
```
Yes, to configure multi hop you have to specify `multihop`.

rzalamena commented:
Rebase to fix merge conflict.
